File: 1219-path-with-maximum-gold/1219-path-with-maximum-gold.py
class Solution:
    def getMaximumGold(self, grid: List[List[int]]) -> int:
        rows,cols=len(grid),len(grid[0])
        def dfs(x,y,res):
            if x<0 or x>=rows or y<0 or y>=cols or grid[x][y]==0:
                return res
            temp=grid[x][y]
            res+=temp
            grid[x][y]=0
            ans=max(dfs(x+1,y,res),dfs(x-1,y,res),dfs(x,y+1,res),dfs(x,y-1,res))
            grid[x][y]=temp
            return ans
        val=0
        for i in range(rows):
            for j in range(cols):
                if grid[i][j]!=0:
                    val=max(val,dfs(i,j,0))
        return val

File: 1219-path-with-maximum-gold/1219-path-with-maximum-gold.cpp
class Solution {
    int rows,cols;
    int val=0;
    void dfs(int x,int y,int res,vector<vector<int>>& grid){
        if(x<0 || x>=rows || y<0 || y>=cols || grid[x][y]==0){
            return ;
        }
        int temp=grid[x][y];
        res+=temp;
        val=max(val,res);
        grid[x][y]=0;
        dfs(x+1,y,res,grid);
        dfs(x-1,y,res,grid);
        dfs(x,y+1,res,grid);
        dfs(x,y-1,res,grid);
        res-=temp;
        grid[x][y]=temp;
    }
public:
    int getMaximumGold(vector<vector<int>>& grid) {
        rows=grid.size();
        cols=grid[0].size();
        for(int i=0;i<rows;i++){
            for(int j=0;j<cols;j++){
                if(grid[i][j]!=0){
                    dfs(i,j,0,grid);
                }
            }
        }
        return val;
    }
};